### PR TITLE
Adds the block hierarchy navigation menu to the header

### LINKED ISSUE #9212: Using dispatch( 'core/editor' ).selectBlock for a nested block focuses on its inner child instead
**Describe the bug**
We are creating UI for being able to select nested blocks from a list of links. Added a new component which displays a root-level block and its inner first level children in a list of links. Clicking on a link triggers `dispatch( 'core/editor' ).selectBlock( clientId )` with the relevant block's `clientId`. For some reason instead of the block with the specified `clientId` being selected its first inner child gets selected.

For example if the blocks are nested A > B > C then calling `dispatch( 'core/editor' ).selectBlock( clientId )` with the `clientId` of `B` focuses on the block `C` is instead. Debugging `clientId` on [these lines](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/reducer.js#L644-L646) shows that for some reason this runs twice -- first time with the correct `clientId` and the second time with the `clientId` of the first inner block (`C`).

Note that using `selectBlock` with the root level element (`A` in the example) selects the correct block.

PR and the line in the PR using `selectBlock`: https://github.com/Automattic/amp-wp/pull/1347/files#diff-34cc40d7d5ae6b1abcceb3757490f8ecR45

**Expected behavior**
The block which `clientId` is used with `dispatch( 'core/editor' ).selectBlock` should be selected.

**Additional context**
- Gutenberg version: 3.6.2

### LINKED ISSUE #9628: Enhance the tools for selecting child/parent blocks
Selecting the parent of a block is not quite as easy as it should be. For some blocks, such as Columns, it can be fiddly to select the right block to apply changes to. Selecting the inner block might be easy, but due to overlapping outlines, selecting the parent might not be. 

Let's explore ways to make this easy and obvious so that Gutenberg can scale to more complex blocks with additional inner blocks. 

## Breadcrumbs

The first improvement we could make, could be breadcrumbs. In the top toolbar we'd always show the selected block, or the selected block and its parent:

model a blockquote with passthrough prop

☝️ Note: This mockup assumes the blockquote becomes a container for inner blocks, which it isn't yet. But in this case, it illustrates that we have selected a _Paragraph_ inside a _Quote_ block. You can click the Quote icon to select the quote. 

If the block doesn't have any inner blocks, it's just a block type indicator, which has been requested many times. 

In order to be an unobtrusive interface, yet still be there when you need it, we'd only ever show _two levels of nesting_ here. For example if you had selected a paragraph inside a quote inside a columns block, we'd only show the quote and the paragraph. Clicking the quote would then change the breadcrumb to show the columns block and the quote block. 

## Click-through

For more complicated blocks, such as the Columns block, we should look to existing desktop apps and mobile solutions for patterns to emulate. One consistent pattern is to group things together and require you to _drill down into the group_ in order to manipulate contents. This is a pattern you might see in Illustrator or Sketch, where you double-click to enter a group. Or in MacOS where when you enter Overview mode, you see previews of all open apps, but you have to select an app before you can manipulate its contents.

It is also already, in a way, a pattern we employ in Gutenberg, where _the selected block can show additional controls_. 

Here is a Cover Image hovered. Note how even though the Heading block inside is hovered, it's the Cover Image block that's highlighted.

model b 01 cover image hovered

Here, the Cover Image is selected. Now inner blocks are made manipulable:

model b 02 cover image selected

When you select an inner block, it is now selected. The parent block is still shown because both are part of a group. 

model b 04 cover image child selected

## Complementary Features

This would be two ways to make it easy to select parent and/or child blocks with simplicity and confidence. 

The features would be designed to work together. For example we might want to enable the "click-through" by default on all blocks that have inner blocks, but provide an optional "passthrough" property which blocks could declare if they felt they could do without the click-through tools. 

For example when the blockquote receives innerblocks, we'd probably want to disable the clickthrough for that block as it's rare that you need to select the quote itself, and when you need to do that, the breadcrumb might be sufficient. 

We might also find that the Cover Image used in these mockups works well as is, without the need for a clickthrough. But it is very likely that the clickthrough will make managing the Columns block that much easier — one click to select the columns block and apply an alignment, another click to select an inner block. 

## Next Steps

Your thoughts are welcome.

We will also want to explore additional features for complex blocks, such as slideshows where an inner block might be out of view depending on the implementation of said block.

But as a start, these two features might make a positive impact. Already today, the "clickthrough" is implemented on mobile, so it would be a matter of refining that, and expanding it beyond that breakpoint.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #9628 and closes #9212

This is still WIP, there are some things we need to polish before landing this:

- [x] The design is ready.
- [x] Add some unit tests for the new selectors and an e2e test for the navigation
- [x] Fix https://github.com/WordPress/gutenberg/issues/9212 which is causing the inner blocks to be selected when we select the parent.

Doesn't mean we can't start testing and reviewing the PR though.

### CODE REVIEW

**Comment by mtias** [comment]
"Styles" should probably be displayed at the top.

**Comment by mtias** [comment]
This is working quite well for me. Agreed on needing to fix #9212. We should also update the single column block icon.

**Comment by mtias** [comment]
Currently:

image

**Comment by youknowriad** [comment]
I fixed the selection bug, I added some tests. The technical pieces should be ready. We might want to do some design tweaks now (cc @jasmussen ) before landing.

**Comment by youknowriad** [comment]
Oh I think @jasmussen is AFK this week (sorry for the pings), I'll see what we can do here :).

**Comment by youknowriad** [comment]
I tried tweaking the design as much as I can. Let me know how I should polish it further. and yes, this is ready for a final review.

**Comment by Soean** [comment]
Currently:
bildschirmfoto 2018-10-15 um 13 13 54

**Comment by mtias** [comment]
> I tried tweaking the design as much as I can. Let me know how I should polish it further.

Let's try to compress it a little so that the second level doesn't come so far into the right.

**Comment by youknowriad** [comment]
screen shot 2018-10-15 at 12 37 55

Compressed, here's a multi-level tree

**Comment by mtias** [comment]
Looking good, updated the icon.

**Comment by mtias** [comment]
Final thing: can we use a background on the "selected" block element?

image

**Comment by tofumatt** [comment]
Yes please to the selected block.

**Comment by youknowriad** [comment]
Updated

screen shot 2018-10-15 at 13 49 14

**Comment by mtias** [comment]
Might need to use a darker gray for the text on the selected item.

**Comment by youknowriad** [comment]
What about this?

screen shot 2018-10-15 at 13 55 20

**Review by mtias — APPROVED** [review]

File: packages/editor/src/store/selectors.js
Comment (aduth) [inline packages/editor/src/store/selectors.js]: This in combination with `getBlockRootClientId` isn't very optimized, particularly when a consuming UI is always present on the screen. Mentioning because I've seen it come up already as one of the higher time-wasters in performance audit results. I might think we should try to do one of:

- Avoid calling the selector if possible (e.g. only when dropdown open)
- Memoize one or both selectors
- Reshape state to optimize upward-path traversal
   - This one is actually fairly straight-forward with adding a `rootClientId` property to the block state. I've only been reluctant because it's technically redundant, but I think performance would win out here.

File: packages/editor/src/components/block-list/block.js
Comment (aduth) [inline packages/editor/src/components/block-list/block.js]: This is a future bug waiting to happen.

File: packages/editor/src/components/block-list/block.js
Comment (aduth) [inline packages/editor/src/components/block-list/block.js]: What is it we're trying to do here? And why is it inconsistent with the behavior in `componentDidMount`?

Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: Okay, so if it's to avoid "select" actually causing one of its inner blocks to become selected, can't this also happen in a newly-mounted block as well?

Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: Maybe we should reconsider whether to call `focusTabbable` on a selection update as well. I don't recall what we need it for.

**Comment by youknowriad** [comment]
> Also: would love to see how this works with a really big block tree. What's the scrolling situation?

It works pretty well since the popover is designed to adapt to the screen and scroll if necessary.

> Is that enough contrast for a11y?

It's waiting for a review

File: packages/editor/src/components/block-list/block.js
Comment (youknowriad) [reply packages/editor/src/components/block-list/block.js]: How would you solve it?

Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: I worked my way back to questioning whether we need it at all in following comments. We can revisit if still then necessary.

File: packages/editor/src/components/block-list/block.js
Comment (youknowriad) [reply packages/editor/src/components/block-list/block.js]: > Maybe we should reconsider whether to call focusTabbable on a selection update as well. I don't recall what we need it for.

Initial implementation here was to avoid selecting innerBlocks in all cases but this breaks the writing flow when inserting columns (it actually broke 2 e2e tests) because the first paragraph was not focused in that case. I did consider whether it's fine to not focus the first paragraph in that case as assume we hit "tab" or click no the appender of the column block first but I think from a "Writing flow" perspective, it's better if we focus the first paragraph in a column directly when creating a columns block.

File: packages/editor/src/components/block-list/block.js
Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: We have an `isInSameBlock` function in `editor/utils/dom.js` which could work well for this.

**Comment by chrisvanpatten** [comment]
@youknowriad Thanks, missed that tag :)

File: packages/editor/src/components/block-list/block.js
Comment (youknowriad) [reply packages/editor/src/components/block-list/block.js]: @aduth It won't work for the column's block appender. As it's in the same block but we don't want to focus it.

File: packages/editor/src/components/block-list/block.js
Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: >it's better if we focus the first paragraph in a column directly when creating a columns block.

Personally I've found this behavior to be a weird / unexpected side effect of focus behaviors. On the Layout<->Writing spectrum, Columns falls more on the former, so I don't necessarily see the paragraph creation / focus being a necessity.

Comment (youknowriad) [reply packages/editor/src/components/block-list/block.js]: > As "Columns" evolve, it'd be natural to focus on the container when it is added rather than on the paragraph within.

Yes one idea I had was to rewrite the columns block similarly to the table block where you select the number of columns before typing but I though this was out of the scope of this PR anyway.

**Review by karmatosed — APPROVED** [review]
Approving design. Thanks for this.

File: packages/editor/src/components/block-list/block.js
Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: I'm getting a sense we're becoming very ad hoc with how and when we expect focus transitions to occur upon block selection, hacking in various conditions to preserve intended flows while losing sight of what those intended flows are. It's signalling to me that naively checking for / comparing on `isSelected` from `BlockListBlock` is not serving us very well, and that perhaps it ought to be moved / refactored to more appropriately capture the user interactions where we expect it to occur: as best I can tell, this is mostly in splitting / merging / replacing actions occurring while typing within a `RichText`.

**Review by tofumatt — DISMISSED** [review]
So this is really cool and keyboard navigation *within* the modal works great. You can use <kbd>Space</kbd> or <kbd>Enter</kbd> to activate a block and navigate to it. Super cool and makes for *excellent* block nav. Adding nicer focus styles could be cool, if they are the right contrast.

**However**, there's not actually a way, that I can see, to *access* the toolbar item with the keyboard. Because right now it's contextual (it only shows for blocks with `innerBlocks`, and only shows when they're activated), I can't tab to it. Tabbing to it also takes a long time. A keyboard shortcut would be good, but I also think this should always be visible and should show just root blocks in order.

If that's the case we could name it "Block Navigation" and we could address #5694.

So I think:

1. this should have a keyboard shortcut
2. it should always appear, not contextually appear
3. it should show all blocks

I hope that's not a pain to implement, but if we did it it'd be soooooooooo good for keyboard users.

**Comment by tofumatt** [comment]
Oh, I forgot to mention: it passes colour contrast.

screenshot 2018-10-15 at 20 10 37

👍

File: packages/editor/src/components/block-list/block.js
Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: Seems like a problem there is that we consider the default appender as a valid text field to target? For what reason is it `<input type="text">`? Should we exclude `readOnly` from auto-focus?

Comment (aduth) [reply packages/editor/src/components/block-list/block.js]: I'd be content if we at least abstract this to some common utility in `editor/utils/dom.js`. The file was created after this same sort of bugginess crept up repeatedly, to have a single point of updates for DOM inspection functions.

e.g. `isInSameRootClientId`

File: packages/editor/src/components/block-navigation/dropdown.js
Comment (gziolo) [inline packages/editor/src/components/block-navigation/dropdown.js]: Can you use `Path` component instead of HTML tag? We probably should refactor all code occurrences and add linter rule for that.

File: packages/editor/src/components/block-list/block.js
Comment (youknowriad) [reply packages/editor/src/components/block-list/block.js]: The appender is an input to allow the writing flow navigation (we discussed this at some point :) I think).

I extracted to a dom utility.

File: packages/editor/src/components/block-navigation/dropdown.js
Comment (aduth) [reply packages/editor/src/components/block-navigation/dropdown.js]: >add linter rule for that.

I did the hard work:

https://astexplorer.net/#/gist/f7f5d60c5bbc381708bff5a035cfe1ad/beb956688752e5b3cf616e77c7b153130b3b979b

Comment (youknowriad) [reply packages/editor/src/components/block-navigation/dropdown.js]: Thanks Andrew, 

screen shot 2018-10-16 at 15 57 52

I'll open a separate PR, as there's still some work to do :P

**Comment by aduth** [comment]
>However, there's not actually a way, that I can see, to access the toolbar item with the keyboard. 

It is possible with the <kbd>Ctrl + `</kbd> region navigation:

https://cloudup.com/cSZe9jH9eWJ

File: packages/editor/src/components/block-navigation/dropdown.js
Comment (youknowriad) [reply packages/editor/src/components/block-navigation/dropdown.js]: Should we also use the `G` cc @gziolo ?

Comment (youknowriad) [reply packages/editor/src/components/block-navigation/dropdown.js]: What about rect...?

**Comment by mtias** [comment]
We cannot use it in the browser, but `<kbd>Command + T</kbd>` would be nice as navigation.

File: packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js
Comment (tofumatt) [inline packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js]: Turns out this keyboard command might be a bit too fast for some users. I noticed to get it to happen I had to press <kbd>Esc</kbd> twice in a row *pretty quickly*. I'm not if that's an OS-level setting (I'm guessing it's something our keyboard shortcut handler defines), but at its current speed it's *way* too tight a tolerance for users with mobility issues.

We should probably find another shortcut for this. I know there are so few, but we'll have to 🤷‍♂️

I'm open to suggestions and I'll try to come up with one.

**Review by tofumatt — DISMISSED** [review]
I updated VMWare earlier and it meant I had to do a dance to get my Windows VM with NVDA up again. Sorry 'bout that.

Overall this works really well.

The main issue here is around the keyboard shortcut. I tried using a screenreader and things worked well, including with the keyboard shortcut ("Block Navigation List" is spoken when opened with the keyboard shortcut, which is good.)

I think we need a modifier key shortcut. What do we have available?

**Review by tofumatt — COMMENTED** [review]
Arg, sorry, I was wrong about the keyboard shortcut giving context to the list, but I have a fix for it incoming.

File: packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js
Comment (tofumatt) [reply packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js]: Possible keyboard options I've tried thus far:

* <del><kbd>⌥⌘B</kbd></del> (Used by Chrome for Bookmarks Bar)
* <del><kbd>⇧⌘B</kbd></del> (Used by Chrome for Bookmarks Page)

File: packages/editor/src/components/block-navigation/index.js
Comment (tofumatt) [inline packages/editor/src/components/block-navigation/index.js]: Oh, @youknowriad pinged me and pointed out this shouldn't be here; we should move this into the block as content (an off-screen header I guess).

I used it with NVDA this way and it only read the label when I used the keyboard shortcut, not when I activated the dropdown using the toggle button. The toggle button provides context as to what this is so it seemed like noise to add it to the content.

But if it's not too annoying to read it out again after the button announcing the content we can move it in, sure 😄

**Comment by afercia** [comment]
> We cannot use it in the browser, but Command + T would be nice as navigation.

> I think we need a modifier key shortcut. What do we have available?

As discussed at length in https://github.com/WordPress/gutenberg/pull/3084 and especially in https://github.com/WordPress/gutenberg/issues/3218 single-modifier shortcuts are basically impossible to use because they will very likely conflict with other shortcuts. Also, not all keyboard layouts provide the same keys. A more reliable approach would be using two modifiers. However, the only final solution to this issue is a mechanism to allow users to re-map and customize shortcuts as proposed in https://github.com/WordPress/gutenberg/issues/3218 and as most of the desktop editors do.

**Comment by afercia** [comment]
@tofumatt looking at this PR for the first time, there's a technical mistake that needs to be fixed.
Based on the ARIA specification, elements with role=menu must have direct children with role=menuitem.

This menu instead contains a list, list items contain a divs with role=presentation, which contain buttons with a role=menuitem. 

Ideally, the UL, LI, and DIV should be completely removed. Why are they even used? If it's for styling purposes, then it's not a valid reason to break the accessibility tree. CSS should be used instead of redundant markup.

Less preferably, and it would potentially expose to AT bugs, all the redundant elements need to be removed from the accessibility tree with role=presentation in order to "flatten" the tree and make AT think the buttons are children of the menu.

Currently, the tree is broken and anyone can just check it in the Chrome of Firefox dev tools:

screen shot 2018-10-17 at 00 18 31

**Review by tofumatt — COMMENTED** [review]
I have some changes to make and then I think this is good to go. I'll approve after I push my changes; if you want to check them out and make sure they're alright please do!

File: packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js
Comment (talldan) [reply packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js]: I'll just mention what I posted in slack here. I started compiling a list of browser shortcuts in a spreadsheet a while ago:
https://docs.google.com/spreadsheets/d/1nK1frKawxV7aboWOJbbslbIqBGoLY7gqKvfwqENj2yE/edit#gid=0

I note that we had access( 'l' ) earmarked for Align Left. How about access( 'o' ), as in O for outline?

**Comment by tofumatt** [comment]
The only thing to note here is NVDA reads out "Block Navigation Block Navigation menu" when this is opened. I wonder if the `label` prop on `MenuGroup` is repeating the label somehow?

That is a bug but we can tweak it after we merge this (which we should do for UI freeze). Once this is merged either file a follow-up issue and ping me or I'll do it 😄

**Comment by tofumatt** [comment]
Changed the shortcut to "o" (see: https://wordpress.slack.com/archives/C02QB2JS7/p1539717395000100) and tweaked the tests. This gets the 👍 from me.

**Review by tofumatt — APPROVED** [review]
Changed the shortcut to "o" (see: https://wordpress.slack.com/archives/C02QB2JS7/p1539717395000100) and tweaked the tests. This gets the 👍 from me.

**Comment by afercia** [comment]
Shortcuts: for future reference, worth nothing on the new WAI-ARIA Authoring Practices 1.1 (still not an official WCAG recommendation) there's a pretty good section dedicated to keyboard shortcuts. It's more about what _not_ to do: https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_shortcuts_assigning

**Comment by talldan** [comment]
I've also made a PR updating the docs with that new keyboard shortcut, would be great if someone could approve:
https://github.com/WordPress/gutenberg/pull/10677

**Comment by talldan** [comment]
One other thing, I noticed this when testing the new menu #10679. I don't think it's a massive issue, but it did cause some weirdness.

**Comment by paaljoachim** [comment]
I am testing 4.1 RC2
I am not totally read up on how this works but I noticed the above wireframe with the icon beside Content Structure. Opening..
screen shot 2018-10-22 at 17 09 31

I see it is a way to move around in the document. I also expected to be able to move the items directly in the list. Instead of having to go to the document and move items there.

**Comment by tofumatt** [comment]
@simbalion and @paaljoachim: Interesting ideas about expectations from these features. I actually think they'd go together; an outline of the selected block would be a great addition. And then a mover would make sense because you could see what you're moving.

Not sure about the mover but the outline I'd definitely 👍 as an idea. Could you both file new issues for the features you'd like? This PR is merged so it's not the best place to discuss additional features, but new issues would be great. Thanks!

**Comment by simbalion** [comment]
Moving blocks would work presumably the same as it normally does, the only difference would be the visibility of each block's borders, and the padding between them. Blocks could be quickly and visibly arranged, and then the 'explode' status could be toggled off to return to WYSIWYG editing.

**Comment by paaljoachim** [comment]
It seems this issue: https://github.com/WordPress/gutenberg/issues/7115 with Insert Summary feature is somewhat similar to the Block Navigation.

**Pull request merged**
